caput/memh5.py

```diff
@@ -2152,40 +2187,89 @@ def _map_json(value):
 
 
 def deep_group_copy(
-    g1, g2, convert_dataset_strings=False, convert_attribute_strings=False
+    g1,
+    g2,
+    selections=None,
+    convert_dataset_strings=False,
+    convert_attribute_strings=False,
 ):
-    """Copy full data tree from one group to another."""
+    """
+    Copy full data tree from one group to another.
+
+    Copies from g1 to g2. An axis downselection can be specified by supplying the
+    parameter 'selections'. If for example 'g1' is
+
+    ```
+    foo:
+       bar: [1, 2, 3]
+    data: [0.1, 0.2, 0.3]
+    ```
+    , calling
+    ```
+    deep_group_copy(g1, g2, {"foo": {"bar": slice(2)}})
+    ```
+    would select the first two indexes in g1["foo"]["bar"], resulting in the
+    following 'g2':
+
+    ```
+    foo:
+       bar: [0,1]
+    data: [0.1, 0.3]
+    ```
+
+    Parameters
+    ----------
+    g1 : h5py.Group
+        Deep copy from this group.
+    g2 : h5py.Group
+        Deep copy to this group.
+    selections : dict
+        If this is not None, it should have a subset of the same hierarchical structure
+        as g1, but ultimately describe axis selections for group entries as valid
+        numpy indexes.
+    convert_strings : bool, optional
+        Convert string attributes (or lists/arrays of them) to ensure that they are
+        unicode.
+    """
+
+    if selections is None:
+        selections = slice(None)
 
     copyattrs(g1.attrs, g2.attrs, convert_strings=convert_attribute_strings)
 
     # Sort to ensure consistent insertion order
     for key in sorted(g1):
         entry = g1[key]
+        if isinstance(selections, dict):
+            selection = selections.get(key, slice(None))
+        else:
```

Comment:
This one.

Reply:
Removing this broke the test. We need to iterate through the items, in case `selections` is a dict.
But because `deep_group_copy` also calls itself recursively, it might not be a dict.